### PR TITLE
CAF-2670: Changes to enable support for resetting the 'trackTo' field to null when a job has no sub tasks.

### PR DESCRIPTION
**Pull request by mulhollandc**
https://jira.autonomy.com/browse/CAF-2670

I have introduced changes to the WF to support the completion of job with no subtasks.
This change allows the tracking 'trackTo' field to be set to null. The batch worker will utilise these changes.

Associated PRs for CAF-2670:
Worker-Batch: https://github.com/JobService/worker-batch/pull/6
Worker-GlobFilter: https://github.com/JobService/worker-globfilter/pull/4
Job-Service: https://github.com/JobService/job-service/pull/32

### CODE REVIEW

**Review by glucy — APPROVED** [review]

File: worker-api/src/main/java/com/hpe/caf/api/worker/WorkerResponse.java [hunk 1]
```diff
@@ -40,6 +41,22 @@
...
+     *
+     * @param queue the reference to the queue that the response data should be put upon. This can be null if no queue is provided
+     * @param status the status of the message the Worker is returning
+     * @param data the serialised task-specific data returned from the Worker internals
+     * @param msgType the task-specific message classifier
+     * @param version the task-specific message API version
+     * @param context the new context to add to the task message, can be null
+     * @param resetTrackTo indicates if the tracking 'trackTo' field should be reset
+     */
+    public WorkerResponse(final String queue, final TaskStatus status, final byte[] data, final String msgType, final int version, final byte[] context, boolean resetTrackTo)
```
Comment (dermot-hardy) [inline worker-api/src/main/java/com/hpe/caf/api/worker/WorkerResponse.java]: New parameter should be marked `final`.

File: worker-core/src/main/java/com/hpe/caf/worker/core/WorkerTaskImpl.java [hunk 2]
```diff
@@ -143,11 +143,26 @@ private TaskMessage createResponseMessage(final boolean includeTaskContext, fina
...
+        //  Check if the 'trackTo' field needs to be reset.
+        final TrackingInfo trackingInfo;
+        if (response.getResetTrackTo()) {
+            //  Reset trackTo field to null;
+            final TrackingInfo taskMessageTracking = taskMessage.getTracking();
+            trackingInfo = new TrackingInfo(taskMessageTracking.getJobTaskId(),
+                    taskMessageTracking.getStatusCheckTime(),
+                    taskMessageTracking.getStatusCheckUrl(),
+                    taskMessageTracking.getTrackingPipe(),
+                    null);
```
Comment (dermot-hardy) [inline worker-core/src/main/java/com/hpe/caf/worker/core/WorkerTaskImpl.java]: - Don't copy all the fields - instead use the copy constructor and then call the `setTrackTo()` mutator to change that one field.  For an example see the `updateTaskId()` method.
- You call `taskMessage.getTracking()` and then use it without checking if it is `null`.  I suspect this will cause a `NullPointerException` for messages that haven't been submitted using the Job Service.

File: worker-caf/src/main/java/com/hpe/caf/worker/AbstractWorker.java [hunk 3]
```diff
@@ -130,9 +130,9 @@ protected final WorkerResponse createSuccessResult(final V result, final byte[]
      *
      * @return a WorkerResponse that represents a success
      */
-    protected final WorkerResponse createSuccessNoOutputToQueue()
+    protected final WorkerResponse createSuccessNoOutputToQueue(boolean resetTrackTo)
```
Comment (dermot-hardy) [inline worker-caf/src/main/java/com/hpe/caf/worker/AbstractWorker.java]: This looks like a breaking change - any worker previously using `createSuccessNoOutputToQueue()` would presumably not work following this change.  Better to add a new method for this.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]
I'm still musing over some of these changes.  I want another review round after this one.

**Comment by mulhollandc** [comment]
Hi @dermot-hardy Thanks for the feedback. I have made the changes.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]
Does this code definitely work?  Looking in `checkForTrackingTermination()` in `WorkerCore.java` it looks like it would cause a `NullPointerException`.  (It is hard to follow the Worker Framework code though so I might have missed something.)

File: worker-core/src/main/java/com/hpe/caf/worker/core/WorkerTaskImpl.java [hunk 4]
```diff
@@ -143,11 +143,24 @@ private TaskMessage createResponseMessage(final boolean includeTaskContext, fina
 
         final String responseMessageType = response.getMessageType();
 
+        //  Check if the 'trackTo' field needs to be reset.
+        final TrackingInfo trackingInfo;
+        if (response.getResetTrackTo()) {
+            //  Reset trackTo field to null;
+            trackingInfo = taskMessage.getTracking();
+            if (trackingInfo != null) {
+                trackingInfo.setTrackTo(null);
```
Comment (dermot-hardy) [inline worker-core/src/main/java/com/hpe/caf/worker/core/WorkerTaskImpl.java]: My previous comment was to use the copy constructor and then call the `setTrackTo()` mutator - you are just using the mutator without calling the copy constructor first.  The copy constructor is here: https://github.com/WorkerFramework/worker-framework/blob/9548a0ed0a2963fd3805430dff5277644c43cb5b/worker-api/src/main/java/com/hpe/caf/api/worker/TrackingInfo.java#L80

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: worker-api/src/main/java/com/hpe/caf/api/worker/WorkerResponse.java [hunk 5]
```diff
@@ -40,6 +41,22 @@
...
+     *
+     * @param queue the reference to the queue that the response data should be put upon. This can be null if no queue is provided
+     * @param status the status of the message the Worker is returning
+     * @param data the serialised task-specific data returned from the Worker internals
+     * @param msgType the task-specific message classifier
+     * @param version the task-specific message API version
+     * @param context the new context to add to the task message, can be null
+     * @param resetTrackTo indicates if the tracking 'trackTo' field should be reset
+     */
+    public WorkerResponse(final String queue, final TaskStatus status, final byte[] data, final String msgType, final int version, final byte[] context, final boolean resetTrackTo)
```
Comment (dermot-hardy) [inline worker-api/src/main/java/com/hpe/caf/api/worker/WorkerResponse.java]: I think this should be more generic.  Rather than having a `resetTrackTo` parameter that sets the `trackTo` pipe to `null` just allow the `trackTo` pipe to be set.  In principle I can think of cases where you'd want to set this to a value other than `null`.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

**Comment by mulhollandc** [comment]
Hi @dermot-hardy Additional changes addressed.

File: worker-caf/src/main/java/com/hpe/caf/worker/AbstractWorker.java [hunk 6]
```diff
@@ -132,7 +132,18 @@ protected final WorkerResponse createSuccessResult(final V result, final byte[]
...
+        return createSuccessNoOutputToQueue("");
+    }
+
+    /**
+     * Utility method for creating a WorkerReponse that represents a success, but does not send a message to the worker's output message.
+     *
+     * @param trackTo the tracking 'trackTo' pipe to set
+     * @return a WorkerResponse that represents a success
+     */
+    protected final WorkerResponse createSuccessNoOutputToQueue(final String trackTo)
```
Comment (dermot-hardy) [inline worker-caf/src/main/java/com/hpe/caf/worker/AbstractWorker.java]: This method doesn't make sense, because it hardcodes no target queue but allows a final tracking queue to be specified.  Replace it with a `createTaskCompleteResponse()` method.

File: worker-core/src/main/java/com/hpe/caf/worker/core/WorkerTaskImpl.java [hunk 7]
```diff
@@ -143,17 +143,44 @@ private TaskMessage createResponseMessage(final boolean includeTaskContext, fina
 
         final String responseMessageType = response.getMessageType();
 
+        //  Check if a tracking change is required.
+        final TrackingInfo trackingInfo;
+        if (null != response.getTrackTo()) {
```
Comment (dermot-hardy) [inline worker-core/src/main/java/com/hpe/caf/worker/core/WorkerTaskImpl.java]: I believe this code can be simplified by checking for empty string first rather than `null`.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

**Comment by mulhollandc** [comment]
Hi @dermot-hardy Hopefully final changes here. I have implemented what I hope we agreed on yesterday.

**Review by dermot-hardy — APPROVED** [review]
